docs/gnetlist.1.in
.TH gnetlist 1 "@DATE@" "gEDA Project" @VERSION@
.SH NAME
gnetlist - gEDA/gaf Netlist Extraction and Generation
.SH SYNOPSIS
.B gnetlist
[\fIOPTION\fR...] [\fB-p\fR \fIBACKEND\fR|\fB-g\fR \fIBACKEND\fR] [--] \fIFILE\fR...

.SH DESCRIPTION
.PP

\fBgnetlist\fR is a netlist extraction and generation tool and is part
of the gEDA (GPL Electronic Design Automation) toolset.  It takes one
or multiple electronic schematics as input and outputs a netlist.  A
netlist is a machine-interpretable description of the way that
components in an electronic circuit are connected together and is
commonly used as the input to a PCB layout program such as
\fBpcb\fR(1) or to a simulator such as \fBgnucap\fR(1).

A normal \fBgnetlist\fR run is carried out in two steps.  First, the
\fBgnetlist\fR frontend loads the specified human-readable schematic
\fIFILE\fRs and compiles them to an in-memory netlist description.
Then, a \fIBACKEND\fR is used to export the connection and component
data to one of many supported netlist formats.

\fBgnetlist\fR is extensible in Python and Scheme.

.SH GENERAL OPTIONS
.TP 8
\fB-q\fR
Quiet mode. Turns off all warnings/notes/messages.
.TP 8
\fB-v\fR, \fB--verbose\fR
Verbose mode.  Output all diagnostic information.
.TP 8
\fB-L\fR \fIDIRECTORY\fR
Prepend \fIDIRECTORY\fR to the list of directories to be searched for
Python backends, Scheme backends, and other Scheme files.
.TP 8
\fB-p\fR \fIBACKEND\fR
Specify the Python netlist backend to be used.
.TP 8
\fB-g\fR \fIBACKEND\fR
Specify the Scheme netlist backend to be used.
.TP 8
\fB-O\fR \fISTRING\fR
Pass an option string to the backend.
.TP 8
\fB--list-backends\fR
Print a list of available netlist backends.  A backend name is
prefixed with the letters `P' and `G' in brackets if it is available
as a Python and as a Scheme backend, respectively.
.TP 8
\fB-o\fR \fIFILE\fR
Specify the filename for the generated netlist.  By default, output is
directed to `output.net'.
If `-' is given instead of a filename, the output is directed to the
standard output.
.TP 8
\fB-l\fR \fIFILE\fR
Specify a Scheme file to be loaded before the backend is loaded or
executed.  This option can be specified multiple times.
.TP 8
\fB-m\fR \fIFILE\fR
Specify a Scheme file to be loaded between loading the backend and
executing it.  This option can be specified multiple times.
.TP 8
\fB-c\fR \fIEXPR\fR
Specify a Scheme expression to be executed during \fBgnetlist\fR
startup.  This option can be specified multiple times.
.TP 8
\fB-i\fR
After the schematic files have been loaded and the netlist generated,
enter an interactive Python interpreter.  The netlist structure is
available as `netlist' in the local dictionary.
.TP 8
\fB--report-gui\fR
Report warnings and errors in a GUI dialog.
.TP 8
\fB-h\fR, \fB--help\fR
Print a help message.
.TP 8
\fB-V\fR, \fB--version\fR
Print \fBgnetlist\fR version information.
.TP 8
\fB--\fR
Treat all remaining arguments as schematic filenames.  Use this if you
have a schematic filename which begins with `-'.

.SH BACKENDS
.PP
Currently, \fBgnetlist\fR includes the following backends:

.TP 8
\fBallegro\fR
Allegro netlist format.
.TP 8
\fBbae\fR
Bartels Autoengineer netlist format.
.TP 8
\fBbom\fR, \fBbom2\fR
Bill of materials generation.
.TP 8
\fBcalay\fR
Calay netlist format.
.TP 8
\fBcascade\fR
RF Cascade netlist format.
.TP 8
\fBdrc\fR, \fBdrc2\fR
Design rule checkers (\fBdrc2\fR is recommended).
.TP 8
\fBdumpindent\fR, \fBdumpjson\fR, \fBdumplihata\fR, \fBdumptext\fR, \fBdumpxml\fR
Dump the netlist in various generic formats.
.TP 8
\fBeagle\fR
Eagle netlist format.
.TP 8
\fBewnet\fR
Netlist format for National Instruments ULTIboard layout tool.
.TP 8
\fBfuturenet2\fR
Futurenet2 netlist format.
.TP 8
\fBgeda\fR
Internal gEDA netlist format (used for testing and diagnostics).
.TP 8
\fBgossip\fR
Gossip netlist format.
.TP 8
\fBgsch2pcb\fR
Backend used for \fBpcb\fR(1) file layout generation by
\fBgsch2pcb\fR(1).  It is not recommended to use this backend
directly.
.TP 8
\fBguile\fR
Compatibility layer for old backends written in Scheme.
.TP 8
\fBliquidpcb\fR
LiquidPCB netlist format.
.TP 8
\fBmakedpend\fR
Auto-generate dependencies for Makefiles.
.TP 8
\fBmathematica\fR
Netlister for analytical circuit solving using Mathematica.
.TP 8
\fBmaxascii\fR
MAXASCII netlist format.
.TP 8
\fBosmond\fR
Osmond netlist format.
.TP 8
\fBpads\fR
PADS netlist format.
.TP 8
\fBpartslist1\fR, \fBpartslist2\fR, \fBpartslist3\fR
Bill of materials generation backends (alternatives to \fBbom\fR and
\fBbom2\fR).
.TP 8
\fBPCB\fR
\fBpcb\fR(1) netlist format.
.TP 8
\fBpcbfwd\fR
Internally used by \fBpcb\fR(1)'s "Import Schematics" mechanism.
.TP 8
\fBpcbpins\fR
Generates a \fBpcb\fR(1) action file for forward annotating pin/pad
names from schematic to layout.
.TP 8
\fBprotelII\fR
Protel II netlist format.
.TP 8
\fBredac\fR
RACAL-REDAC netlist format.
.TP 8
\fBspice\fR, \fBspice-sdb\fR, \fBspice-noqsi\fR
SPICE-compatible netlist format (\fBspice-sdb\fR and \fBspice-noqsi\fR
are recommended).
Suitable for use with \fBgnucap\fR(1).
.TP 8
\fBswitcap\fR
SWITCAP switched capacitor simulator netlist format.
.TP 8
\fBsystemc\fR
Structural SystemC code generation.
.TP 8
\fBtango\fR
Tango netlist format.
.TP 8
\fBtedax\fR
tEDAx netlist format (used by \fBpcb-rnd\fR(1)).
.TP 8
\fBvams\fR
VHDL-AMS code generation.
.TP 8
\fBverilog\fR
Verilog code generation.
.TP 8
\fBvhdl\fR
VHDL code generation.
.TP 8
\fBvipec\fR
ViPEC Network Analyser netlist format.


.SH EXAMPLES
.PP
These examples assume that you have a schematic file `stack_1.sch' in
the current directory.
.PP
\fBgnetlist\fR requires that at least one schematic and either a
Python backend, a Scheme backend, or interactive mode are specified on
the command line.  If a backend name is specified with `\-p
\fIBACKEND\fR' (for a Python backend) or `\-g \fIBACKEND\fR' (for a
Scheme backend), \fBgnetlist\fR will output a netlist:

.nf
	gnetlist \-p dumptext stack_1.sch
.ad b

.PP
The netlist output will be written to a file called `output.net'
in the current working directory.

.PP
You can specify the output filename by using the `\-o' option:

.nf
	gnetlist \-p dumptext \-o /tmp/stack.netlist stack_1.sch
.ad b

.PP
Output will now be directed to `/tmp/stack.netlist'.

.PP
You could run (for example) the `spice-sdb' backend against the
schematic if you specified `\-p spice-sdb', or you could generate a
bill of materials for the schematic using `\-p partslist1'.

.PP
To obtain a Python prompt to run Python commands directly, you can
use the `\-i' option.

.nf
	gnetlist \-i stack_1.sch
.ad b

.PP
\fBgnetlist\fR will load `stack_1.sh' and then enter an interactive
Python interpreter.  The netlist structure is available as `netlist'
in the local dictionary.

.SH ENVIRONMENT
.TP 8
.B GEDADATA
specifies the search directory for Scheme and rc files.  The default
is `${prefix}/share/gEDA'.
.TP 8
.B GEDADATARC
specifies the search directory for rc files.  The default is `$GEDADATA'.

.SH AUTHORS
See the `AUTHORS' file included with this program.

.SH COPYRIGHT
.nf
Copyright \(co 1999-@YEAR@ gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
\fBgschem\fR(1), \fBgsymcheck\fR(1), \fBpcb\fR(1), \fBpcb-rnd\fR(1), \fBgnucap\fR(1)


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2019 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.


Notable changes in gEDA/gaf 1.10.0
==================================

General changes
---------------

* Added the library "xornstorage" and the Python package "gaf" as a
  (currently quite minimal) way to access gEDA/gaf functionality from
  outside of gschem.  These are on the long term intended to replace
  libgeda, making the core gEDA/gaf functionality available as a
  library to both gschem and other applications.

* Added command-line tool `xorn' for low-level schematic and symbol
  file manipulation.

* Added experimental XML file format (currently only supported by
  gnetlist and the `xorn' command line utility).

* Objects colors are limited to 21 (up to and including FREESTYLE4).
  More recent colors added for GUI purposes can be set in colormaps
  but aren't automatically added to the file format specification.

* The symbol directories "gnetman", "verilog", and "vhdl" are now
  installed to the new location PREFIX/share/gEDA/extra-sym/.  This
  allows adding the default library with "component-library-search".

* The directory to which gEDA programs write log transcripts can be
  changed by setting the environment variable `GEDALOG' (the default
  being ~/.gEDA/logs/).

* All tests now use the parallel test harness, speeding up `make
  check' runs considerably when used with multiple jobs (-jN).

* Fixed bugs:
  - lp-698501   gschem file preview function renders files twice
  - lp-698524   gnetlist ignores duplicate inout-symbols
  - lp-698768   No ordering requirement for attribute attach/detach
  - lp-701637   gschem "Scheme Interaction" window
  - lp-1266316  gschem exits on color value < 0
  - lp-1367462  undo messes with the view even though undo panzoom is disabled
  - lp-1492834  These docked dialogs should remember their size
  - lp-1496899  gschem: dock window wishlist
  - lp-1497659  gnetlist -v -g spice-sdb generates bad verbose netlist
  - lp-1514209  gschem: multiattrib Value textview improvements
  - lp-1532794  Build fails on missing colors.h
  - lp-1576918  Memory and file handle leak in libgeda
  - lp-1590757  partlist3 yields wrong netlist in some cases
  - lp-1625474  xorn does not build on FreeBSD
  - lp-1661961  "format not a string literal and no format arguments" warning
  ...and a lot more which never made it to the bugtracker.


gschem changes
--------------

* Tool windows are now "dockable", i.e., they can be either used as
  modal/non-modal dialogs or docked to the edges of the main window.

* The action mechanism used for menu items, tool buttons, and key
  bindings has been redesigned, making several improvements possbile:

  - Actions can now be used interchangeably in the main menu, toolbar,
    context menu, and keymap.  They are defined in a single place
    ("actions.c" for C actions and "builtins.scm" for Guile actions),
    including metadata like name, icon, and tooltip.

  - Menu items and tool buttons representing options have a little
    check or radio box beside them indicating their current state.

  - Menu items and tool buttons are greyed out while they can't be
    used.

  - Toolbar and context menu are customizable.

  - Menu and toolbar definitions are simple Scheme variables, allowing
    users to add actions to the menu without having to copy the menu
    definition.  Actions are represented in Scheme code as applicable
    SMOBs, allowing them to be called like procedures while retaining
    the action information necessary for rendering the menus/toolbar.

  - "Repeat Last Action" (usually bound to ".") uses the same logic as
    the middle mouse button repeat action does, i.e., only actions
    that "make sense" qualify for repeating.

* The menus, toolbar, and keymap have been revamped to more closely
  resemble the "standards" for a desktop application.  Some actions
  have been assigned more accessible or memorable keys.

* Added a command-line option "--control-fd=stdin|FD" which allows
  remote-controlling gschem via a file descriptor.  This is intended
  for project managers like Igor2's "genxproj".

* File operations have been consolidated: opening, saving, closing a
  file and so on should now behave identical regardless of how the
  operation was initiated.  Related improvements:

  - When opening multiple files, the first one (instead of the last
    one) is initially shown.

  - The symversion= mismatch dialog is shown whenever a symversion=
    mismatch is detected while opening a file, not just for files
    specified on the command line.

  - gschem now keeps track of which files are untitled, listing them
    as "untitled page" instead of assigning them a temporary filename.
    The Scheme procedure `page-filename' returns #f for untitled
    pages.

  - "Save all" used to produce spurious "untitled_1.sch" files.  This
    was due to three issues working together, which have been fixed:

    - When gschem is invoked without specifying filenames on the
      command line, an untitled page is created.  This page wasn't
      removed when opening files, staying invisibly in the background.

    - "Save all" didn't check which files needed saving but just
      unconditionally saved anything.

    - "Save all" didn't behave like the interactive "Save", asking for
      a filename for untitled pages, but just saved each page to its
      internally stored filename.

  - The file chooser dialog on "Open" and "Save As" is now always
    opened in the current directory.

  - When opening a file which is already open but has since been
    changed on disk, the user is asked whether to reload the file.

  - When the currently edited file has been modified on disk, a

gnetlist-legacy/src/gnetlist.c
/* gEDA - GPL Electronic Design Automation
 * gnetlist - gEDA Netlist
 * Copyright (C) 1998-2010 Ales Hvezda
 * Copyright (C) 1998-2019 gEDA Contributors (see ChangeLog for details)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include <config.h>
#include <locale.h>
#include <version.h>

#include <stdio.h>
#include <sys/param.h>
#include <sys/types.h>
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#include <dirent.h>
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif


#include <libgeda/libgeda.h>
#include <libgeda/libgedaguile.h>

#include "../include/globals.h"
#include "../include/prototype.h"
#include "../include/gettext.h"

void gnetlist_quit(void)
{
    s_clib_free();
    s_slib_free();
    s_rename_destroy_all();
    /* o_text_freeallfonts(); */

    /* Free GSList *backend_params */
    g_slist_free (backend_params);

    g_slist_free (input_files);
}


/* \brief Print a list of available backends.
 * \par Function Description
 * Prints a list of available gnetlist backends by searching for files
 * in each of the directories in the current Guile %load-path.  A file
 * is considered to be a gnetlist backend if its basename begins with
 * "gnet-" and ends with ".scm".
 *
 * \param pr_current  Current #TOPLEVEL structure.
 */
void
gnetlist_backends (TOPLEVEL *pr_current)
{
  SCM s_load_path;
  GList *backend_names = NULL, *iter = NULL;

  /* Look up the current Guile %load-path */
  s_load_path = scm_variable_ref (scm_c_lookup ("%load-path"));

  for ( ; !scm_is_null (s_load_path); s_load_path = scm_cdr (s_load_path)) {
    SCM s_dir_name = scm_car (s_load_path);
    char *dir_name;
    DIR *dptr;
    struct dirent *dentry;

    /* Get directory name from Scheme */
    g_assert (scm_is_true (scm_list_p (s_load_path))); /* Sanity check */
    g_assert (scm_is_string (scm_car (s_load_path))); /* Sanity check */
    dir_name = scm_to_utf8_string (s_dir_name);

    /* Open directory */
    dptr = opendir (dir_name);
    if (dptr == NULL) {
      g_warning (_("Can't open directory %s: %s\n"),
                 dir_name, strerror (errno));
      continue;
    }
    free (dir_name);

    while (1) {
      char *name;

      dentry = readdir (dptr);
      if (dentry == NULL) break;

      /* Check that filename has the right format to be a gnetlist
       * backend */
      if (!(g_str_has_prefix (dentry->d_name, "gnet-")
            && g_str_has_suffix (dentry->d_name, ".scm")))
        continue;

      /* Copy filename and remove prefix & suffix.  Add to list of
       * backend names. */
      name = g_strdup (dentry->d_name + 5);
      name[strlen(name)-4] = '\0';
      backend_names = g_list_prepend (backend_names, name);
    }

    /* Close directory */
    closedir (dptr);
  }

  /* Sort the list of backends */
  backend_names = g_list_sort (backend_names, (GCompareFunc) strcmp);

  printf (_("List of available backends: \n\n"));

  for (iter = backend_names; iter != NULL; iter = g_list_next (iter)) {
    printf ("%s\n", (char *) iter->data);
  }
  printf ("\n");

  scm_remember_upto_here_1 (s_load_path);
}


void main_prog(void *closure, int argc, char *argv[])
{
    int i;
    int argv_index;
    char *cwd;
    gchar *str;
    gchar *filename;

    TOPLEVEL *pr_current;

    /* set default output filename */
    output_filename = g_strdup("output.net");

    argv_index = parse_commandline(argc, argv);
    cwd = g_get_current_dir();

    scm_set_program_arguments (argc, argv, NULL);

    /* this is a kludge to make sure that spice mode gets set */
    /*  Hacked by SDB to allow spice netlisters of arbitrary name
     *        as long as they begin with "spice".  For example, this spice
     *  netlister is valid: "spice-sdb".
     */
    if (guile_proc) {
        if (strncmp(guile_proc, "spice", 5) == 0) {
            netlist_mode = SPICE;
        }
    }

    libgeda_init();

    /* create log file right away */
    /* even if logging is enabled */
    s_log_init ("gnetlist");

    s_log_message(_(
        "gEDA/gnetlist-legacy version %s%s.%s\n"
        "gEDA/gnetlist-legacy comes with ABSOLUTELY NO WARRANTY; see COPYING for more details.\n"
        "This is free software, and you are welcome to redistribute it under certain\n"
        "conditions; please see the COPYING file for more details.\n\n"),
        PREPEND_VERSION_STRING, PACKAGE_DOTTED_VERSION, PACKAGE_DATE_VERSION);

#if defined(__MINGW32__) && defined(DEBUG)
    fprintf(stderr, _("This is the MINGW32 port.\n\n"));
#endif

    /* register guile (scheme) functions */
    g_register_funcs();

    scm_dynwind_begin (0);
    pr_current = s_toplevel_new ();
    edascm_dynwind_toplevel (pr_current);

    /* Evaluate Scheme expressions that need to be run before rc files
     * are loaded. */
    scm_eval (pre_rc_list, scm_current_module ());

    g_rc_parse (pr_current, argv[0], "gnetlistrc", rc_filename);
    /* immediately setup user params */
    i_vars_set (pr_current);

    s_rename_init();

    if(list_backends) {
      gnetlist_backends(pr_current);
      exit (0);
    }

    /* Evaluate the first set of Scheme expressions before we load any
     * schematic files */
    scm_eval (pre_backend_list, scm_current_module ());

    i = argv_index;
    while (argv[i] != NULL) {
      GError *err = NULL;

      if (g_path_is_absolute(argv[i])) {
        /* Path is already absolute so no need to do any concat of cwd */
        filename = g_strdup (argv[i]);
      } else {
        filename = g_build_filename (cwd, argv[i], NULL);
      }

      if (!quiet_mode) {
        s_log_message (_("Loading schematic [%s]\n"), filename);
        fprintf (stderr, _("Loading schematic [%s]\n"), filename);
      }

      s_page_goto (pr_current, s_page_new (pr_current, filename));

      if (!f_open (pr_current, pr_current->page_current, filename, &err)) {
        g_warning ("%s\n", err->message);
        fprintf (stderr, _("ERROR: Failed to load '%s': %s\n"),
                 filename, err->message);
        g_error_free (err);
	exit(2);
      }

      /* collect input filenames for backend use */
      input_files = g_slist_append(input_files, argv[i]);

      i++;
      g_free (filename);
    }

    /* Change back to the directory where we started.  This is done */
    /* since gnetlist is a command line utility and will deposit its output */
    /* in the current directory.  Having the output go to a different */
    /* directory will confuse the user (confused me, at first). */
    if (chdir (cwd) == -1) {
      fprintf (stderr,
               _("ERROR: Failed to restore working directory to `%s': %s\n"),
               cwd, g_strerror (errno));
      exit (1);
    }
    /* free(cwd); - Defered; see below */

    if (argv[argv_index] == NULL) {
        fprintf (stderr, _(
            "ERROR: No schematics files specified for processing.\n"
            "\nRun `%s --help' for more information.\n"), argv[0]);
        exit (1);
    }

#if DEBUG
    s_page_print_all(pr_current);
#endif

    /* Load basic gnetlist functions */
    scm_primitive_load_path (scm_from_utf8_string ("gnetlist.scm"));

    if (guile_proc) {
      SCM s_backend_path;

      /* Search for backend scm file in load path */
      str = g_strdup_printf("gnet-%s.scm", guile_proc);
      s_backend_path = scm_sys_search_load_path (scm_from_locale_string (str));
      g_free (str);

      /* If it couldn't be found, fail. */
      if (scm_is_false (s_backend_path)) {
        fprintf (stderr, _(
            "ERROR: Could not find backend `%s' in load path.\n"
            "\nRun `%s --list-backends' for a full list of available backends.\n"),
            guile_proc, argv[0]);
        exit (1);
      }

      /* Load backend code. */
      scm_primitive_load (s_backend_path);

      /* Evaluate second set of Scheme expressions. */
      scm_eval (post_backend_list, scm_current_module ());
    }

    s_traverse_init();
    s_traverse_start(pr_current);

    /* Change back to the directory where we started AGAIN.  This is done */
    /* because the s_traverse functions can change the Current Working Directory. */
    if (chdir (cwd) == -1) {
      fprintf (stderr,
               _("ERROR: Failed to restore working directory to `%s': %s\n"),
               cwd, g_strerror (errno));
      exit (1);
    }
    g_free(cwd);

    /* Run post-traverse code. */
    scm_primitive_load_path (scm_from_utf8_string ("gnetlist-post.scm"));

    if (interactive_mode) {
        scm_c_eval_string ("(set-repl-prompt! \"gnetlist-legacy> \")");
        scm_shell (0, NULL);
    } else if (guile_proc) {
        /* check size here hack */
        str = g_strdup_printf ("(%s \"%s\")", guile_proc, output_filename);
        scm_c_eval_string (str);
        g_free (str);
        /* gh_eval_str_with_stack_saving_handler (input_str); */
    } else {
        fprintf(stderr,
                _("You gave neither backend to execute nor interactive mode!\n"));
    }

    gnetlist_quit();

    scm_dynwind_end();
}

int main(int argc, char *argv[])
{
#if ENABLE_NLS
    setlocale (LC_ALL, "");
    setlocale (LC_NUMERIC, "C");
    bindtextdomain ("geda-gnetlist-legacy", LOCALEDIR);
    textdomain ("geda-gnetlist-legacy");
    bind_textdomain_codeset("geda-gnetlist-legacy", "UTF-8");
#endif
    scm_boot_guile (argc, argv, main_prog, 0);
    return 0;
}
